src/win/pipe.c

```diff
@@ -942,20 +951,81 @@ int uv_pipe_listen(uv_pipe_t* handle, int backlog, uv_connection_cb cb) {
 }
 
 
+static DWORD WINAPI uv__pipe_zero_read_interrupter(void* param) {
+  uv__pipe_interrupter_param_t* interrupterParam;
+  int r;
+
+  interrupterParam = (uv__pipe_interrupter_param_t*) param;
+  for (;;) {
+    r = WaitForSingleObject(interrupterParam->completedEvent,
+                            PIPE_ZERO_READ_INTERRUPT_INTERVAL);
+    if (r == WAIT_TIMEOUT) {
+      /* ReadFile thread is active - interrupt ReadFile to give other */
+      /* processes a chance to act upon pipe */
+      pCancelSynchronousIo(interrupterParam->thread);
+    } else {
+      /* ReadFile thread has terminated - cleanup the handles and exit */
+      CloseHandle(interrupterParam->completedEvent);
+      CloseHandle(interrupterParam->thread);
+      uv__free(interrupterParam);
+      return 0;
+    }
+  }
+}
+
+static HANDLE uv__start_zero_read_interrupter() {
+  uv__pipe_interrupter_param_t* param;
+
+  param = (uv__pipe_interrupter_param_t*) uv__malloc(
+    sizeof(uv__pipe_interrupter_param_t));
```

Comment:
this should be able to use `alloca` – the application needs to guarantee that this pointer is never accessed after the function returns, to avoid race conditions, anyways.